pkg/storage/azure/azure_test.go

```diff
@@ -329,6 +333,132 @@ func TestConfigEnvWithUserKey(t *testing.T) {
 	}
 }
 
+// custom sender for mocking
+type sender struct {
+	response []*http.Response
+	body     string
+}
+
+// Do implements the Sender interface for mocking
+// Do accepts the passed request and body, then appends the response and emits it.
+func (s *sender) Do(r *http.Request) (*http.Response, error) {
+	resp := &http.Response{
+		StatusCode: http.StatusOK,
+		Request:    r,
+		Body:       io.NopCloser(bytes.NewBufferString(s.body)),
+	}
+	s.response = append(s.response, resp)
+	return resp, nil
+}
+
+func TestUserProvidedTags(t *testing.T) {
+	for _, tt := range []struct {
+		name         string
+		userTags     []configv1.AzureResourceTag
+		expectedTags map[string]string
+		infraName    string
+		responseBody string
+	}{
+		{
+			name:      "no-user-tags",
+			infraName: "some-infra",
+			// only default tags
+			expectedTags: map[string]string{
+				"kubernetes.io_cluster.some-infra": "owned",
+			},
+			responseBody: `{"nameAvailable":true}`,
+		},
+		{
+			name:      "with-user-tags",
+			infraName: "test-infra",
+			userTags: []configv1.AzureResourceTag{
+				{
+					Key:   "tag1",
+					Value: "value1",
+				},
+				{
+					Key:   "tag2",
+					Value: "value2",
+				},
+			},
+			// default tags and user tags
+			expectedTags: map[string]string{
+				"kubernetes.io_cluster.test-infra": "owned",
+				"tag1":                             "value1",
+				"tag2":                             "value2",
+			},
+			responseBody: `{"nameAvailable":true}`,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			sender := &sender{
+				body: tt.responseBody,
+			}
+
+			storageConfig := &imageregistryv1.ImageRegistryConfigStorageAzure{}
+
+			drv := NewDriver(context.Background(), storageConfig, nil)
+			drv.authorizer = autorest.NullAuthorizer{}
+			drv.sender = sender
+
+			_, _, err := drv.assureStorageAccount(
+				&Azure{
+					SubscriptionID: "subscription-id",
+					ResourceGroup:  "resource-group",
+				},
+				&configv1.Infrastructure{
+					Status: configv1.InfrastructureStatus{
+						InfrastructureName: tt.infraName,
+						Platform:           configv1.AzurePlatformType,
+						PlatformStatus: &configv1.PlatformStatus{
+							Type: configv1.AzurePlatformType,
+							Azure: &configv1.AzurePlatformStatus{
+								ResourceTags: tt.userTags,
+							},
+						},
+					},
+				},
+			)
+
+			if err != nil {
+				t.Errorf("unexpected error %q", err)
+			}
+
+			for _, resp := range sender.response {
+				if resp != nil && resp.Request != nil && resp.Request.Body != nil {
+
+					reqBody := make(map[string]interface{})
+					if err := json.NewDecoder(resp.Request.Body).Decode(&reqBody); err != nil {
+						t.Fatalf("error decoding request: %q", err)
+					}
+
+					// ignore reqest without tags
+					if _, ok := reqBody["tags"]; ok {
```

Comment:
What if none of the response.Request.Body contain tags?

Reply:
That's a good point. I have added a check for this. Thanks for noting.